### PR TITLE
Support for mouseenter event for a row

### PR DESCRIPTION
**Pull request by szpetny**
**What kind of change does this PR introduce?** (check one with "x")
- [ ] Bugfix
- [x ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other... Please describe:

**What is the current behavior?** (You can also link to an open issue here)
https://github.com/swimlane/ngx-datatable/issues/909

**What is the new behavior?**
Mouseenter event handler available in body-row.component.

**Does this PR introduce a breaking change?** (check one with "x")
- [ ] Yes
- [ x] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

### CODE REVIEW

**Comment by amcdnl** [comment]
Hmmm, activate event. I think of activate when it has focus via click or keyboard. this almost needs a new event type.

**Comment by szpetny** [comment]
dunno, for me hovering is also kind of activation ಠ_ಠ but anyways I need this feature, so want me to introduce a new emitter?

**Comment by amcdnl** [comment]
Ya, make sense. No I'll go with this.

**Comment by bmcleod** [comment]
Thanks for adding this feature! Is it possible to get an example on how to use it/documentation?

**Comment by szpetny** [comment]
You need to bind an event handler `<ngx-datatable (activate)="onActivate($event)">
        </ngx-datatable>` then inside this event handler do a check `onActivate(event) {
if (event.type === 'mouseenter') {
 console.log(event);
}
  }`,
 cheers

**Comment by wizarrc** [comment]
Just a word of caution, this may break people who don't check the event type but depend somehow on the event to maintain some kind of state. An earlier project I worked on did just that.

@amcdnl might be a good idea to add to the changelog that you check the event type if you need to discriminate against hover activation.

**Comment by wizarrc** [comment]
@szpetny Sorry for being late on the thread. I like the idea of having many forms of activation, including mouse hovering. I can anticipate cool new ways of interacting with the table. Interesting idea!

I'd be curious from a performance point of view. Imagine for a moment that you are scrolling via the scroll wheel on the mouse. Since the activate event is being listened to on the angular zone, it will trigger a change detection upon handling the event. Now the question is will that event trigger many times during a scroll, causing performance to suffer? If so, it might be a good idea to register the event outside of angular zone, filter out any events during scrolling operations, then if there are any listeners, handle the event back in the angular zone, with full change detection logic.

Prior to this, css would be the only way to style for hovering not causing any change detection problems.

**Comment by szpetny** [comment]
perhaps there should be some delay to avoid too frequent firing, probably could look at it at the weekend

**Comment by szpetny** [comment]
Ok, I committed sth, not sure if you like it bc it introduces another listener (to mouseleave). 
Anyways this is the fork:
https://github.com/szpetny/ngx-datatable

**Comment by wizarrc** [comment]
@szpetny the problem with counting on the `mouseleave` event is that it doesn't always trigger. An example of this is during debugging, and setting a break point. Another example where it doesn't trigger the event is when the OS displays another window for whatever reason, and the browser looses focus, then when the browser gets focus again, the mouse has already moved thus preventing the `mouseleave` event from triggering.

**Comment by wizarrc** [comment]
@szpetny related issue https://github.com/swimlane/ngx-datatable/issues/989

I think the perf is fine for now though. If you want to make any perf changes, I think something like this would be preferred.

datatable-body.component.ts
```
ngOnInit() {
    this.zone.runOutsideAngular(() => {
      this.element.addEventListener('mouseenter', this.onMouseenter.bind(this));
    });
}
ngOnDestroy() {
  this.element.removeEventListener(this.onMouseEnter.bind(this));
}
```
datatable.component.ts
```
<datatable-body
...
(activate)="onActivate($event)"
...
onActivate(event) {
  if(this.activate.observers.length) {
    this.zone.run(() => this.activate.emit(event));
}
```

**Comment by wizarrc** [comment]
Unfortunately the mere listening to the event on the datatable component `(activate)="onActivate($event)"` will mark change detection to check but will not actually do the detection till next change detection, which may or may not fire, causing a weird inconsistency that could cause hard to find bugs.

An even better approach is to somehow know if there are any listeners in the main datatable component from the body component and only then trigger the event in the body component. That would mean there would have to be yet another `@input()` presenting that data to the body component. That means yet another change detection variable to check during change detection.

FYI, I solved this problem in my fork by gutting out all this state transfers between components and using a service provider between components, but that's a whole other topic. Needless to say, it gets complicated to make perf improvements without restructuring the internals.

**Comment by szpetny** [comment]
Aww, didn't know that mouseleave is unreliable. Then I guess let it be as it is, apparently the straightforward solutions are the best sometimes..

**Comment by wizarrc** [comment]
@szpetny I updated my recommendation in comment https://github.com/swimlane/ngx-datatable/pull/912#issuecomment-331716695

**Comment by wizarrc** [comment]
@szpetny yeah, straightforward can be less buggy, but using the built-in angular constructs can be a perf problem, but to solve the issue, it can get quite complicated, and hard to find all edge cases, as you can see in my spitball example.

**Comment by szpetny** [comment]
Yes, so making a conclusion, you said it might be a performance problem, then you did some tests and said it is actually ok, though it could be more neat, then I checked my new idea and it failed, then you came out with an idea which is sophisticated, but also complicated and can cause hard-to-find bugs. So I understand: let's leave it as it is? At least for a while

**Comment by wizarrc** [comment]
@szpetny I'm pretty sure my latest refresh on my code example solves the hard to find bug and event consistency issue, but it's still hard to follow for those not sure how the angular eventing system, zones, and change detection works internally. I vote to leave it alone till the internals are better structured.
